[Wrik] iP

thewrik opened the pull request:
### Hello world, I am Morty!

I am an upgraded Task Manager who takes after Duke.

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))


Morty frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- **SUPER** FAST to use

All you need to do is,

1. download it from [here]().
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉
5. And it is FREE!

Features:

- [X]  Managing tasks
- [X]  Managing deadlines 
- [ ]  Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the main method:
```

public class Main {
    public static void main(String[] args) {
        Storage.initialise();
        Display.greet();
        Parser.parseInput();
    }
}
```

Commit:
Add Gradle support

thewrik committed:
Level 1 - Greeting Messages

Stylised Greeting Messages, basic echoing capacity added.

thewrik committed:
Level-2 Add and List of Tasks

Created modularity of Tasks, addition and listing functionality added.

thewrik committed:
Level-3 Mark as Done

Mark and Unmark functionality added to each task.

thewrik committed:
Level-4 Support Task Types

Support added for 3 different Task types (ToDos, Events, Deadlines), backwards compatible with Task without type mentioned.

thewrik committed:
A-Enums

Changed commands to Enum structure.

thewrik committed:
Level-5 Error Handling

Handles the following errors:
- Commands outside the scope
- Commands without task names for task addition commands
- Commands without relevant time fields for task additions where required.

thewrik committed:
Level-6 Delete

Deletion Support added, Implementation changed to use factory methods.

thewrik committed:
Refractor

thewrik committed:
Level-7 Save functionality

Save file option added through Storage class, tasklist autosaved after every update, save command support added.

thewrik committed:
Level-8 Basic Datetime Support added

Uses LocalDate type, handles dd-mm-yyyy formats exclusives, for event and deadline tasks.

thewrik committed:
Level-8 Basic Datetime Support added

Uses LocalDate type, handles dd-mm-yyyy formats exclusives, for event and deadline tasks.

thewrik committed:
Level-8 DateTime

Date print format changed.

thewrik committed:
Level-8 DateTime

Date print format changed.

thewrik committed:
Level-8 DateTime

Date print format changed.

thewrik committed:
Revert "Level-8 Basic Datetime Support added"

This reverts commit 47a4ab9c6cc4644dc9dfe6d448db5ceae63b1b9e.

thewrik committed:
Revert "Level-8 DateTime"

This reverts commit faaff1b2a2d78d76dfc00bac73c93f3765954cda.

thewrik committed:
Merge branch 'branch-Level-7'

thewrik committed:
Merge branch 'branch-Level-8'

thewrik committed:
fix merge conflicts, imports

thewrik committed:
Revert "Level-8 DateTime"

This reverts commit 52233ba2e7bc1c6274e05bcee30b8a83ce31f545.

thewrik committed:
Merge branch 'branch-Level-8'

thewrik committed:
A-MoreOOP Refractoring and Organising

thewrik committed:
A-Junit Tests added

Testing added to Task Manager, supporting 4 key functions.

thewrik committed:
JAR Creation changes

thewrik committed:
Level-9 Find Functionality Added

Find functionality added through Stream API

thewrik committed:
A-CodingStandard

config file added, changes made accordingly

thewrik committed:
A-JavaDoc added

JavaDoc added

thewrik committed:
Merge branch 'branch-A-CodingStandard'

thewrik committed:
Merge branch 'branch-A-JavaDoc'

jxt00 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,14 @@
+import duke.ui.Display;
+import duke.ui.Parser;
+
+public class Duke {
```
I feel that it would be better if header comments for public classes/methods were included. I noticed the same issue in several other places too.

jxt00 commented on src/main/java/duke/commands/Commands.java:
```diff
@@ -0,0 +1,13 @@
+package duke.commands;
+
+public enum Commands {
+    todo,
+    deadline,
+    event,
+    list,
+    mark,
+    unmark,
+    bye,
+    delete,
+    find,
```
Should the enum type's fields be written in uppercase?

yusufaine replied:
Same as what @\jxt00 mentioned, enums are conventionally upper-cased.

jxt00 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,40 @@
+    public static void initialise() {
+        try {
+            Path storageFolderPath = Paths.get("data");
+            if (Files.notExists(storageFolderPath)) Files.createDirectory(storageFolderPath);
+            Path storageFilePath = storageFolderPath.resolve(Paths.get("duke.tasks.txt"));
+            if (Files.notExists(storageFilePath)) Files.createFile(storageFilePath);
```
I feel the loop body should be wrapped by curly brackets irrespective of how many lines there are in the body.

zzhikai replied:
Like what @jxt00 mentioned, do consider using curly brackets for the if statement, I think it will improve readability and adhere to the coding standards.
```suggestion
            if (Files.notExists(storageFolderPath)) {
                Files.createDirectory(storageFolderPath);
            }
            Path storageFilePath = storageFolderPath.resolve(Paths.get("duke.tasks.txt"));
            if (Files.notExists(storageFilePath)) {
                Files.createFile(storageFilePath);
            }
```

yusufaine replied:
+1 to the wrapping of the if-statement, I was also guilty of committing this, but it's part of the module's convention, and I think it does help with readability (even though it takes up a bit more space 😆 )

jxt00 commented on src/main/java/duke/tasks/DeadlineTask.java:
```diff
@@ -0,0 +1,38 @@
+        this.deadline = deadline;
+    }
+
+    public static DeadlineTask DeadlineTaskBuilder(String taskName) throws InvalidParameterException {
```
Should the method name be a verb?

yusufaine replied:
I understand the rationale behind the naming, but again, module convention states that method names should be presented as a verb.
```suggestion
    public static DeadlineTask buildDeadLineTask(String taskName) throws InvalidParameterException {
```

jxt00 commented on src/main/java/duke/tasks/TaskManager.java:
```diff
@@ -0,0 +1,67 @@
+        Task addedTask;
+        taskName = taskName.trim();
+
+        if (taskName.startsWith(Commands.todo.toString())) {
```
I like how you used Egyptian braces here! 👍

jxt00 commented on src/main/java/duke/ui/Display.java:
```diff
@@ -0,0 +1,20 @@
+package duke.ui;
+
+public class Display {
+    private static final String hline = "----------------------------------------------";
+
+    static final String greetingMessage = "Hello! I am Morty, Rick's assistant.\nWhat can I do for you, today?";
+    static final String farewellMessage = "Hope you had fun talking to me, and bye!\nHave a nice day ahead!";
```
Maybe the constant names should be all uppercase?

jxt00 commented on src/main/java/duke/commands/Commands.java:
```diff
@@ -0,0 +1,13 @@
+package duke.commands;
+
+public enum Commands {
```
Would it be less confusing if the class name is singular?

jxt00 reviewed (commented):
Overall, I found your code easy to read for the most part except a few places
where the if conditionals were a bit hard to keep track of. I noted a few minor coding standard violations too. I like how you named your classes as nouns and methods as verbs. 👍

zzhikai reviewed (commented):
LGTM other than the comments already pointed out by @jxt00 , do consider adding JavaDocs to help others understand your code better.

yusufaine commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,14 @@
+package duke;
```
I personally like how clean the entry point of the app is 👍  overall the convention (for this mod) is apart from getters and setters, all other public methods should be JavaDoc'd

yusufaine commented on src/main/java/duke/tasks/DeadlineTask.java:
```diff
@@ -0,0 +1,38 @@
+            throw new InvalidParameterException("A Deadline Task must have a task name.\nPlease try again!");
+        }
+        int lastIndexOfBackslash = taskName.lastIndexOf('/');
+        if (lastIndexOfBackslash == -1) {
+            throw new InvalidParameterException("A Deadline Task must have an associated deadline.\nPlease try again!");
+        }
```
Ayyy nice, you indented this if-block. I think you might've missed out the one earlier in `Storage` 👍

yusufaine commented on src/main/java/duke/tasks/EventTask.java:
```diff
@@ -0,0 +1,39 @@
+        this.eventTime = eventTime;
+    }
+
+    public static EventTask EventTaskBuilder(String taskName) throws InvalidParameterException {
```
Similar to the comment made in your `DeadlineTask`, it should probably be in the form of a verb (even though this one's quite self-explanatory 🤷‍♂️ 

```suggestion
    public static EventTask buildEventTask(String taskName) throws InvalidParameterException {
```

yusufaine commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,26 @@
+
+    public Task(String taskName, String taskType) {
+        this.taskName = taskName;
+        this.isDone = Boolean.FALSE;
```
Kind of curious as to why you didn't set this line as:
```suggestion
        this.isDone = false;
```
But this is more nit-picky then anything.

yusufaine commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,79 @@
+        return String.format("Great, have added the following task for you:\n%s\nNow you have %s.", task, countTasks());
+    }
+
+    static Task get(int index) {
+        return taskList.get(index - 1);
+    }
```
I think `index` should already have been `entryNumber - 1`, this _might_ cause some confusion but it's really just renaming it as:
```suggestion
    static Task get(int entryNumber) {
        return taskList.get(entryNumber - 1);
    }
```

yusufaine commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,79 @@
+
+        prompt.append(String.format("You have %s in the list\n", countTasks()));
+
+        for (Task t : taskList) {
+            prompt.append(String.format("%d. %s\n", task_no++, t.toString()));
+        }
```
I like using enhanced for-loops and `forEach` to iterate through my Collections as well, but since you have to a variable perhaps consider the _boring_ for-loop? 😆 

```suggestion
        for (int i = 0; i < taskList.size(); i++) {
            prompt.append(String.format("%d. %s\n", i+1, t.toString()));
        }
```

At least this way you can get rid of line 29

yusufaine commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    public static String find(String taskKeyword) {
+        List<Task> filteredTasks = taskList.stream()
+                .filter(
+                        task -> Arrays.stream(task
+                                .toString()
+                                .toLowerCase()
+                                .split(" "))
+                                .anyMatch(word -> word.equals(taskKeyword)))
+                .collect(Collectors.toList());
```
I really like this approach! Maybe you can modifying it to (unless I'm misunderstanding what you're attempting to do here):

```suggestion
        List<Task> filteredTasks = taskList
                .stream()
                .filter(task -> task.contains(taskKeyword)
                .collect(Collectors.toList());
```

Lambdas weren't covered in the convention but my TA said to refer to [Google's](https://google.github.io/styleguide/javaguide.html). Typically there is a linebreak between each lambda operation

thewrik replied:
Hey, @yusufaine going to take the other suggestions under advisement, not replying individually to all, but thanks, great observations! 

With regards to this particular comment, I wanted to root out false flags. 
Let the `tasklist` be `[borrow book, read book, booking meetings]`
In my approach, `find("book")` returns `borrow book` and `read book` but not `booking meetings`. I thought of this while mistakenly triggering a search with the query `b`, wasn't pretty. 

Hope that helps.

yusufaine commented on src/main/java/duke/ui/Parser.java:
```diff
@@ -0,0 +1,48 @@
+        if (inputMessage.trim().equals(Commands.list.toString())) {
+            Display.printStylised(TaskManager.list());
+        } else if (inputMessage.startsWith(Commands.mark.toString())) {
+            Display.printStylised(TaskManager.mark(Integer.parseInt(inputMessage.substring(Commands.mark.toString().length() + 1))));
```
This is quite the eye-sore (on git PR view at least) consider storing some of that as a local variable

yusufaine commented on src/test/java/duke/tasks/TaskManagerTest.java:
```diff
@@ -0,0 +1,63 @@
+package duke.tasks;
```
Just wanted to point out that the best-practice states that there's only one `assert*` per method (but I did it the same way as you too 🤣 )

yusufaine commented on .gitignore:
```diff
@@ -15,3 +15,4 @@ bin/
 
```
I'm not sure if you should be including the docs' html files, but I got a shock when I saw that I had to look through 113 files!

yusufaine reviewed (commented):
Overall clean code and rather intuitive to understand, only some slight nit-picks and code convention issues that might've slipped past your checks, hopefully my comments were enough to point them out 👍 

Also -- I'm not sure if you should be including the docs' html files, but I got a shock when I saw that I had to look through 113 files!

thewrik committed:
Merge branch 'add-gradle-support'

thewrik committed:
Checkstyle

thewrik committed:
Level-10 GUI added

thewrik committed:
Update TaskList.java

thewrik committed:
Extension C-BetterSearch is implemented

findBetter added

thewrik committed:
Extension C-BetterSearch

findBetter command: case insensitive complete and partial matching for tasks

The findBetter command is case insensitive and is more user-friendly since one can find tasks which contain the keyword within a word (e.g the keyword "book" would match with tasks containing either of "booking", "bookie", and "booker") which is an improvement from the find function implemented earlier.

thewrik committed:
Extension C-BetterSearch

findBetter command: case insensitive complete and partial matching for tasks

The findBetter command is case insensitive and is more user-friendly since one can find tasks which contain the keyword within a word (e.g the keyword "book" would match with tasks containing either of "booking", "bookie", and "booker") which is an improvement from the find function implemented earlier.

thewrik committed:
Merge branch 'master' of https://github.com/thewrik/ip

thewrik committed:
Level-10 GUI Added

thewrik committed:
Merge branch 'branch-Level-10'

thewrik committed:
Set theme jekyll-theme-minimal

thewrik committed:
docs cleanup

thewrik committed:
A-UserGuide

thewrik committed:
readme typo

thewrik committed:
A-Assertions

thewrik committed:
Merge branch 'heads/A-Assertions'

thewrik committed:
Revert "Merge branch 'heads/A-Assertions'"

This reverts commit 35f7e409d85af78089dcb4cb1ccccb3a1b8622ac, reversing
changes made to 26594c06b3b3a9182f57a4e9d1909f01de6cd175.

thewrik committed:
Merge branch 'master' into Level-10-GUI

thewrik committed:
Merge pull request #2 from thewrik/Level-10-GUI

Level-10 GUI added

thewrik committed:
typo Fixed, Assertions added

thewrik committed:
Merge pull request #3 from thewrik/branch-A-Assertions

typo Fixed, Assertions added

thewrik committed:
Ui.png Added

thewrik committed:
Merge branch 'master' of https://github.com/thewrik/ip

thewrik committed:
Update README.md

thewrik committed:
JavaDocs updated

thewrik committed:
Merge branch 'master' of https://github.com/thewrik/ip

thewrik committed:
Update README.md

thewrik committed:
Update README.md

thewrik committed:
save functionality ammended

thewrik committed:
Update Parser.java

Find Better changed
